File: companion/ZippieCompanionApp/Design/RelayScreen.swift
import NetworkExtension
import SwiftUI
import ZippieCompanionKit

private struct RouterSSIDRow: Identifiable {
    let id = UUID()
    var name: String
}

/// Relay, rebuilt on the design system.
///
/// WHAT CHANGED AND WHY. The old screen was five stock `List` sections in
/// declaration order: background, settings, status, foreground, wiring. That
/// order is the order the features were BUILT in, not the order anyone needs
/// them. Opening it, the first thing you met was a control; the thing you
/// actually came for - is this phone contributing - was three sections down in
/// grey `LabeledContent`.
///
/// It now opens with the state, then the single control that changes it, then
/// the numbers backing it, and only then the endpoint fields and the fallback
/// that almost nobody should touch. Same capabilities, same behaviour, ordered
/// by what the reader needs first.
///
/// The foreground relay is still here and still last. It is the only relay that
/// works on a build whose extension cannot be signed yet (ADR 0020), and
/// deleting a proven capability because it is unglamorous would leave the
/// operator with nothing.
struct RelayScreen: View {
    /// The LIVE mode decision, not a second copy of the reasoning.
    ///
    /// `BondModel` probes the console every five seconds and is the only thing
    /// that knows whether the router answered on its LAN address - which is the
    /// entire basis for contribute versus client (ADR 0022). Before #48 that
    /// verdict was rendered on the Status screen and acted on nowhere, so the
    /// start button below could only ever install a contributor profile.
    @ObservedObject var bond: BondModel

    @StateObject private var tunnel = TunnelController()

    @State private var homeHost = Settings.homeHost
    @State private var homePort = String(Settings.homePort)
    @State private var listenPort = String(Settings.listenPort)
    @State private var routerSSIDRows: [RouterSSIDRow] = {
        let configured = Settings.routerSSIDs
        return (configured.isEmpty ? [""] : configured).map(RouterSSIDRow.init(name:))
    }()
    @State private var currentSSID: String?

    @State private var foregroundRelay: CellularRelay?
    @State private var foregroundStats = CellularRelay.Stats()
    @State private var foregroundRunning = false
    @State private var startError: String?
    @State private var showsAdvanced = false

    private var editingLocked: Bool {
        foregroundRunning || tunnel.status == .connected || tunnel.status == .connecting
    }

    private var running: Bool {
        tunnel.status == .connected || tunnel.status == .connecting
            || tunnel.status == .reasserting
    }

    var body: some View {
        Page {
            header
            control
            numbers
            endpoint
            routerNetworks
            advanced
        }
        .task { await tunnel.refresh() }
        .task {
            while !Task.isCancelled {
                currentSSID = await NetworkFacts.current().ssid
                try? await Task.sleep(nanoseconds: 5_000_000_000)
            }
        }
        .task {
            // Polled because there is no cross-process notification when the
            // extension writes its report. One second is well inside the
            // staleness threshold, so a dead extension shows up within a couple
            // of ticks rather than never.
            while !Task.isCancelled {
                await tunnel.refreshReport()
                // Supervision reads the report this call just refreshed, so it
                // rides the same tick rather than owning a timer that could
                // judge a snapshot one second out of date. Its own thresholds
                // are minutes wide; the polling rate is not the thing that
                // decides anything (see RelaySupervision).
                await tunnel.supervise()
                try? await Task.sleep(nanoseconds: 1_000_000_000)
            }
        }
    }

    // MARK: - state, first

    private var header: some View {
        VStack(alignment: .leading, spacing: Space.snug) {
            Text(verdict.headline)
                .font(Kind.display())
                .tracking(-0.8)
                .foregroundStyle(Ink.primary)
                .fixedSize(horizontal: false, vertical: true)
            // Named when known, so "the router" is never left to be read as
            // the wifi router this phone is joined to (#44 operator
            // follow-up, 2026-08-08). See Settings.routerDisplayName.
            Text(verdict.detail(router: Settings.routerDisplayName))
                .font(Kind.body())
                .foregroundStyle(Ink.secondary)
                .fixedSize(horizontal: false, vertical: true)
            if foregroundRunning {
                // Said here as well as in Advanced, because Advanced is
                // collapsed by default and this caveat decides whether the leg
                // survives the next thirty seconds.
                Note(text: "Running inside the app. Keep this screen open and "
                   + "the phone charging.")
            }
        }
        .padding(.top, Space.section)
        .padding(.bottom, Space.roomy)
        .accessibilityElement(children: .combine)
    }

    /// What the screen may honestly say, decided in the Kit (`RelayVerdict`).
    ///
    /// THE VIEW COMPUTES NO TRUTH. It maps `NEVPNStatus` onto the Kit's
    /// vocabulary and renders the answer. The version that decided for itself
    /// is the one that shipped "Connected to the router, waiting for traffic to
    /// carry." on a phone the router had never dialled (#44, seen live
    /// 2026-08-07) - derived from `cellularReady`, which is a fact about this
    /// phone's own radio and says nothing whatsoever about the far end.
    private var verdict: RelayVerdict {
        if foregroundRunning {
            // In-process counters, so they are current by construction: there
            // is no second binary here that could have died holding them, which
            // is the entire reason RelayStatus carries a timestamp at all.
            return RelayVerdict.evaluate(
                run: .running,
                report: RelayStatus(stats: foregroundStats, updatedAt: Date()))
        }
        let run: RelayRun
        switch tunnel.status {
        case .connected:                run = .running
        case .connecting, .reasserting: run = .starting
        case .disconnecting:            run = .stopping
        default:                        run = .off
        }
        return RelayVerdict.evaluate(run: run, report: tunnel.report)
    }

    // MARK: - the one control

    private var control: some View {
        VStack(alignment: .leading, spacing: Space.snug) {
            if running {
                ActionButton(title: "Stop relaying", role: .destructive) {
                    tunnel.stopTunnel()
                }
            } else {
                ActionButton(title: tunnel.installed ? "Start relaying" : "Install and start",
                             enabled: !foregroundRunning) { startBackground() }
            }

            Note(text: "Runs in a Network Extension, so the leg survives the screen "
               + "locking. iOS shows a VPN badge while it runs; the tunnel carries "
               + "none of this phone's own traffic. Keep the phone charging - "
               + "relaying over cellular runs the radio hot.")

            if let e = tunnel.lastError ?? startError {
                Note(text: e, tone: .warning)
            }
            // WHAT SUPERVISION MADE OF IT, shown only when it found a fault.
            //
            // A relay that is fine has nothing to say here and a permanent
            // "supervision: healthy" row is furniture that trains the eye to
            // skip the place a real sentence would appear - the same reason the
            // Errors row below is hidden at zero. But when the relay IS wedged,
            // the sentence includes why nothing is being done about it, which
            // is the line that stops somebody debugging the router: on
            // 2026-08-22 the router showed 65% loss and a null round trip while
            // this screen said "Ready" and no part of the app explained the
            // gap.
            if let pass = tunnel.supervision, pass.verdict.isFault {
                Note(text: pass.remedy.why, tone: .warning)
            }
        }
    }

    // MARK: - what backs it

    /// The extension's report wins when the tunnel is up; otherwise the
    /// foreground relay's counters. They are never both live - each control
    /// disables the other - because two processes bound to one UDP port is a
    /// bug, not redundancy.
    private var shownStats: CellularRelay.Stats {
        if tunnel.status == .connected, let r = tunnel.report { return r.stats }
        return foregroundStats
    }

    /// "nothing yet" rather than "never": the relay may have started a second
    /// ago, and "never" reads as a verdict on the router when it is a statement
    /// about how long we have been listening.
    private var routerInboundText: String {
        guard let at = shownStats.lastRouterInboundAt else { return "nothing yet" }
        let age = Date().timeIntervalSince(at)
        return age < 2 ? "just now" : "\(RelayVerdict.ago(age)) ago"
    }

    private var numbers: some View {
        VStack(alignment: .leading, spacing: 0) {
            SectionHead(title: "Carried")
            Readout(label: "To home",
                    value: "\(shownStats.upDatagrams) pkt  \(Fmt.bytes(UInt64(max(0, shownStats.upBytes))))")
            Hairline()
            Readout(label: "To router",
                    value: "\(shownStats.downDatagrams) pkt  \(Fmt.bytes(UInt64(max(0, shownStats.downBytes))))")
            Hairline()
            // THE EVIDENCE BEHIND THE HEADLINE, shown rather than implied. Every
            // other row here is a fact about this phone; this is the only one
            // that says the router exists, and #44 was a headline that claimed
            // it without a row like this to back it.
            Readout(label: "From router",
                    value: routerInboundText,
                    tone: shownStats.lastRouterInboundAt == nil ? Ink.degraded : Ink.primary)
            Hairline()
            Readout(label: "Cellular",
                    value: shownStats.cellularReady ? "ready" : "not ready",
                    tone: shownStats.cellularReady ? Ink.primary : Ink.down)
            // Shown only when non-zero. A permanent "Errors 0" row is a line of
            // furniture that trains the eye to skip the place a real number
            // would appear.
            if shownStats.errors > 0 {
                Hairline()
                Readout(label: "Errors", value: "\(shownStats.errors)", tone: Ink.degraded)
            }
            if shownStats.rejectedSources > 0 {
                Hairline()
                Readout(label: "Refused senders", value: "\(shownStats.rejectedSources)",
                        tone: Ink.degraded)
            }
            if let e = shownStats.lastError {
                Note(text: e, tone: .warning)
            }
        }
    }

    // MARK: - where it points

    /// The heading that made a reader ask "which router is it connected to?"
    /// (#44). This block is the HOME end - where the relay forwards TO - and it
    /// sat directly under a headline about the router with nothing to separate
    /// them. Only "Listen on" faces the router, so the note says which is which
    /// rather than leaving the reader to infer it from three field names.
    private var endpointNote: String {
        let what = "Where this phone forwards to: the home end, not the router. "
                 + "The router reaches this phone on the listen port below."
        return editingLocked ? what + " Stop the relay to change these." : what
    }

    private var endpoint: some View {
        VStack(alignment: .leading, spacing: 0) {
            SectionHead(title: "Home transport", note: endpointNote)
            FieldRow(label: "Host", text: $homeHost)
            Hairline()
            FieldRow(label: "Port", text: $homePort, keyboard: .numberPad)
            Hairline()
            FieldRow(label: "Listen on", text: $listenPort, keyboard: .numberPad)
            Note(text: "Point the router's leg at this phone's wifi address on port "
               + "\(listenPort).")
        }
        .disabled(editingLocked)
        .opacity(editingLocked ? 0.5 : 1)
    }

    private var routerNetworks: some View {
        let policy = OnDemandPolicy(routerSSIDs: formRouterSSIDs)
        return VStack(alignment: .leading, spacing: 0) {
            SectionHead(
                title: "Router wifi",
                note: "The contributor tunnel starts on any network listed here. "
                    + "Add every wifi name the router broadcasts. Matching is exact."
            )
            ForEach($routerSSIDRows) { $row in
                let index = routerSSIDRows.firstIndex { $0.id == row.id } ?? 0
                if index > 0 { Hairline() }
                HStack(spacing: Space.snug) {
                    FieldRow(label: "Name \(index + 1)", text: $row.name)
                    if routerSSIDRows.count > 1 {
                        Button {
                            routerSSIDRows.removeAll { $0.id == row.id }
                        } label: {
                            Image(systemName: "minus.circle")
                                .foregroundStyle(Ink.degraded)
                        }
                        .accessibilityLabel("Remove wifi name \(index + 1)")
                    }
                }
            }
            Button {
                routerSSIDRows.append(RouterSSIDRow(name: ""))
            } label: {
                Label("Add wifi name", systemImage: "plus.circle")
                    .font(Kind.caption())
                    .foregroundStyle(Ink.live)
                    .padding(.top, Space.tight)
            }
            .buttonStyle(.plain)
            Note(text: policy.explain(currentSSID: currentSSID),
                 tone: policy.shouldRun(onSSID: currentSSID) ? .plain : .warning)
        }
        .disabled(editingLocked)
        .opacity(editingLocked ? 0.5 : 1)
    }

    private var formRouterSSIDs: [String] {
        RelayConfiguration.normalizedRouterSSIDs(routerSSIDRows.map(\.name))
    }

    // MARK: - the part almost nobody should touch

    /// Collapsed by default, and that is not tidying. The foreground relay and
    /// removing the VPN profile are both ways to end up with a phone that looks
    /// configured and carries nothing; putting them behind a deliberate tap
    /// keeps them reachable without offering them.
    private var advanced: some View {
        VStack(alignment: .leading, spacing: 0) {
            SectionHead(title: "Advanced")

            Button {
                withAnimation(.snappy(duration: 0.2)) { showsAdvanced.toggle() }
            } label: {
                HStack {
                    Text(showsAdvanced ? "Hide" : "Show")
                        .font(Kind.body())
                        .foregroundStyle(Ink.live)
                    Spacer()
                    Image(systemName: showsAdvanced ? "chevron.up" : "chevron.down")
                        .font(.footnote)
                        .foregroundStyle(Ink.live)
                }
                .padding(.vertical, Space.snug)
            }
            .buttonStyle(.plain)

            if showsAdvanced {
                VStack(alignment: .leading, spacing: Space.snug) {
                    Hairline()
                    ActionButton(title: foregroundRunning
                                 ? "Stop foreground relay" : "Start foreground relay",
                                 role: .quiet,
                                 enabled: !running) { toggleForeground() }
                    Note(text: "Runs the relay inside the app, as it did before the "
                       + "Network Extension existed. Only useful on a build whose "
                       + "tunnel cannot be signed yet - iOS suspends background apps, "
                       + "so the leg dies when you leave this screen.")

                    if tunnel.installed {
                        ActionButton(title: "Remove VPN configuration",
                                     role: .destructive) {
                            Task { await tunnel.removeConfiguration() }
                        }
                    }
                }
                .transition(.opacity)
            }
        }
    }

    // MARK: - actions (unchanged behaviour)

    /// Reads the form, and refuses rather than substituting a default. A relay
    /// pointed at the wrong host looks identical to a working one from here.
    private func configFromForm() -> RelayConfiguration? {
        guard let hp = RelayConfiguration.port(homePort),
              let lp = RelayConfiguration.port(listenPort),
              !homeHost.trimmingCharacters(in: .whitespaces).isEmpty else { return nil }
        return RelayConfiguration(homeHost: homeHost.trimmingCharacters(in: .whitespaces),
                                  homePort: hp, listenPort: lp,
                                  routerSSIDs: formRouterSSIDs)
    }

    private func startBackground() {
        guard let config = configFromForm() else {
            startError = "Check the host and ports below."
            return
        }
        startError = nil
        persist(config)
        Task {
            // `client: nil` is the honest value, not a placeholder. Nothing can
            // mint a `ClientConfig` until the pairing ceremony exists (#31), so
            // `TunnelPlan.decide` resolves to contribute with the reason
            // `clientModeNotConfigured` and the profile installed is byte for
            // byte the one that ships today.
            //
            // A refusal arrives as `tunnel.lastError`, which the header above
            // already prefers over `startError` - so the plan's own sentence
            // reaches the screen without a second copy of it here.
            await tunnel.startTunnel(with: config, decision: bond.decision, client: nil)
            Observability.tunnelStatus(tunnel.status, error: tunnel.lastError)
        }
    }

    private func persist(_ config: RelayConfiguration) {
        Settings.homeHost = config.homeHost
        Settings.homePort = Int(config.homePort)
        Settings.listenPort = Int(config.listenPort)
        Settings.routerSSIDs = config.routerSSIDs
    }

    private func toggleForeground() {
        if foregroundRunning {
            Task {
                await foregroundRelay?.stop()
                foregroundRelay = nil
                foregroundRunning = false
            }
            return
        }
        guard let config = configFromForm() else {
            startError = "Check the host and ports below."
            return
        }
        persist(config)
        startError = nil

        let r = CellularRelay(config: config.relayConfig)
        foregroundRelay = r
        Task {
            do {
                try await r.start { s in
                    Task { @MainActor in foregroundStats = s }
                    Observability.relayStats(s)
                }
                await MainActor.run { foregroundRunning = true }
            } catch {
                await MainActor.run {
                    startError = "Start failed: \(error.localizedDescription)"
                    foregroundRelay = nil
                }
            }
        }
    }
}

File: companion-android/app/src/main/java/app/zippie/companion/RelayScreen.kt
package app.zippie.companion

import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.OutlinedTextFieldDefaults
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.input.PasswordVisualTransformation
import androidx.compose.ui.unit.dp
import app.zippie.companion.design.Hairline as Rule
import app.zippie.companion.design.Ink
import app.zippie.companion.design.Kind
import app.zippie.companion.design.Space

/**
 * This phone's own half of the bond: whether it is contributing, the single
 * control that changes that, and the fallback nobody should need.
 *
 * ORDERED THE WAY RelayScreen.swift ORDERS ITSELF, and for the same stated
 * reason: "it now opens with the state, then the single control that changes
 * it, then the numbers backing it, and only then the endpoint fields... ordered
 * by what the reader needs first" rather than the order the features were
 * built in. This file WAS that build order - carrier and wifi address printed
 * above the verdict that explains why they matter - moved here verbatim out of
 * StatusScreen.kt (#parity), which is the moment to fix the ordering along with
 * the address.
 *
 * SPLIT OUT OF StatusScreen.kt, not rewritten. Every sentence, every colour
 * rule and every test this content had stays true; only its screen changed.
 * The screen it left behind was the one the operator actually opens (see
 * StatusScreen.kt's own header comment), and it had grown to hold a router
 * write token editor a glance from a dashboard never needed to see.
 */
@Composable
fun RelayScreen(
    state: BondUiState,
    onStartRelay: () -> Unit,
    onStopRelay: () -> Unit,
    onStartClient: () -> Unit,
    onSaveToken: (String) -> Unit,
) {
    Column(
        Modifier
            .verticalScroll(rememberScrollState())
            .padding(horizontal = Space.margin, vertical = Space.roomy),
    ) {
        SectionTitle("This phone")
        RelaySection(state, onStartRelay, onStopRelay, onSaveToken)

        Spacer(Modifier.height(Space.section))
        SectionTitle("Client mode")
        Text(
            // Said plainly rather than hidden behind a disabled button: the
            // tunnel comes up and carries NOTHING until the gomobile datapath
            // exists (#2246), and a button that looks ready would be a lie.
            "Bonding this phone's own links back home needs the gomobile datapath, " +
                "which is not in this build yet (#2246). The tunnel will establish and " +
                "carry nothing.",
            style = Kind.caption,
            color = Ink.secondary,
        )
        Spacer(Modifier.height(8.dp))
        OutlinedButton(onClick = onStartClient, colors = ButtonDefaults.outlinedButtonColors(contentColor = Ink.live)) { Text("Establish the tunnel anyway") }
    }
}

@Composable
private fun RelaySection(
    state: BondUiState,
    onStartRelay: () -> Unit,
    onStopRelay: () -> Unit,
    onSaveToken: (String) -> Unit,
) {
    val report = state.relay
    val stats = report?.stats
    // THE ONE SENTENCE, evidence-gated (RelayVerdict.kt, ported from
    // ZippieCompanionKit/RelayVerdict.swift and quadseven/zippie#44): a claim
    // about the ROUTER is made only once something has actually arrived from
    // it, decided from a TIMESTAMP rather than a datagram count.
    // upDatagrams/downDatagrams never go down, so a count-based rule reads
    // "Carrying" forever after a single packet, including long after the
    // router stopped dialling this phone - that was the defect here.
    // AND the router's half, when this screen already holds it. The leg the
    // console publishes for THIS phone carries `never_handshaked`, and until
    // #281 that fact was rendered in the leg row below while the headline above
    // it said "Carrying" - the same screen contradicting itself for hours
    // during an outage. `isYou` is matched by relay endpoint, not by name, so a
    // relabelled leg still resolves.
    val routerSeesNothing = state.legs.any { it.isYou && it.neverAnswered }
    val verdict = RelayVerdict.evaluate(report, state.nowMs, routerSeesNothing = routerSeesNothing)

    // THE STATE COMES FIRST, matching RelayScreen.swift's `header`. This used
    // to be three sentences below the carrier and wifi-address readout that
    // explains it - which is the exact "opened on a control... the thing you
    // actually came for three sections down" complaint RelayScreen.swift's own
    // header comment makes about ITS previous version.
    val alarmed = verdict is RelayVerdict.NotReporting
    Text(
        verdict.headline,
        style = Kind.body,
        color = if (alarmed) Ink.down else Ink.primary,
    )
    Text(
        // Not yet named on Android - see the note on RelayVerdict.detail(router:)
        // for why there is no display-name source here to pass.
        verdict.detail(),
        style = Kind.caption,
        color = if (alarmed) Ink.down else Ink.secondary,
    )

    Spacer(Modifier.height(Space.base))
    // ONE CONTROL, decided by the verdict already computed above.
    //
    // Both buttons used to render unconditionally, so a phone that was actively
    // carrying still offered "Start contributing" - which reads as "this is not
    // running" on the one screen whose entire job is saying whether it is. The
    // operator reported exactly that. RelayScreen.swift has said `if running`
    // since it was written; this is Android catching up.
    //
    // Off is the only verdict meaning no relay exists (RelayVerdict.Off: "no
    // report exists ... never been started, or just stopped"). Every other
    // verdict - including NotReporting, where the service is alive but wedged -
    // is a state you leave by stopping, so Stop is the honest control there.
    if (verdict is RelayVerdict.Off) {
        Button(
            onClick = onStartRelay,
            colors = ButtonDefaults.buttonColors(containerColor = Ink.live, contentColor = Ink.ground),
        ) { Text("Start contributing", style = Kind.label) }
    } else {
        OutlinedButton(
            onClick = onStopRelay,
            colors = ButtonDefaults.outlinedButtonColors(contentColor = Ink.live),
        ) { Text("Stop contributing", style = Kind.label) }
    }

    // THE NUMBERS BACKING IT, next - matching RelayScreen.swift's ordering.
    // Raw measurements, shown only while the report is live and current - a
    // corpse's byte counts sitting under "not reporting" would look like more
    // evidence when they are the opposite.
    if (stats != null && verdict !is RelayVerdict.Off && verdict !is RelayVerdict.NotReporting) {
        Spacer(Modifier.height(Space.base))
        Text(
            "${Fmt.bytes(stats.upBytes)} up in ${stats.upDatagrams} datagrams, " +
                "${Fmt.bytes(stats.downBytes)} down in ${stats.downDatagrams}.",
            style = Kind.caption,
            color = Ink.secondary,
        )
        Text(
            if (stats.budget.isConfigured) {
                "Budget: ${Fmt.bytes(stats.dayUsedBytes)} today, " +
                    "${Fmt.bytes(stats.monthUsedBytes)} this month."
            } else {
                "No data cap set, so the relay will spend whatever the bond asks for."
            },
            style = Kind.caption,
            color = Ink.secondary,
        )
        if (stats.rejectedSources > 0) {
            Text(
                "Refused ${stats.rejectedSources} datagrams from senders that were not the router.",
                style = Kind.caption,
                color = Ink.secondary,
            )
        }
        stats.budgetExhausted?.let {
            // Not an error colour: the budget working as designed is not a
            // fault, and colouring it red would train the reader to fix it.
            Text("$it (${stats.budgetBlocked} datagrams refused)",
                style = Kind.caption)
        }
        stats.announce?.let {
            // NOT an error colour by default. "No token set" is a configuration
            // fact, and a refusal is the router's answer - only the second is
            // worth alarming about.
            Text(
                it,
                style = Kind.caption,
                color = Ink.secondary,
            )
        }
        stats.lastError?.let {
            Text(it, style = Kind.caption, color = Ink.down)
        }
    }

    // THE ENDPOINT, last before the fallback - matching RelayScreen.swift's
    // `endpoint` section. A carrier name is something iOS cannot report at
    // all, so it is stated here rather than inferred from the router's
    // hand-written leg label.
    Spacer(Modifier.height(Space.base))
    Text(
        state.carrier?.summary?.let { "Carrier: $it" } ?: "Carrier: not reported by the radio",
        style = Kind.caption,
        color = Ink.secondary,
    )
    Text(
        state.localIp?.let { "Wifi address: $it, listening on ${state.listenPort}" }
            ?: "No wifi address, so no leg can be matched to this phone",
        style = Kind.caption,
        color = Ink.secondary,
    )

    AnnounceSettings(state, onSaveToken)
}

/**
 * The one thing this phone needs from a person before it can join the bond.
 *
 * WHY IT IS ON THIS SCREEN AT ALL. Announcing is authenticated - the router
 * answers 401 without the token - and there is no settings screen to put it on.
 * A build with no way to enter the token would be an announce path that can
 * never run, which is worse than not having one: it looks finished.
 *
 * Masked, never echoed back, and never logged. The field starts empty even when
 * a token IS stored, because reading one back onto a screen is how a token ends
 * up in a screenshot of a bug report.
 */
@Composable
private fun AnnounceSettings(state: BondUiState, onSaveToken: (String) -> Unit) {
    var typed by remember { mutableStateOf("") }

    Spacer(Modifier.height(16.dp))
    Text(
        state.legName?.let { "This phone announces itself as \"$it\"." }
            ?: "This phone has not chosen a leg name yet.",
        style = Kind.caption,
        color = Ink.secondary,
    )
    Text(
        if (state.hasAnnounceToken) {
            "A console write token is stored, so this phone can announce itself. " +
                "A token saved while the relay is running takes effect when it is " +
                "next started."
        } else {
            "No console write token, so the router will refuse to add this phone " +
                "as a leg. It is the router's /var/lib/zippie/console_token."
        },
        style = Kind.caption,
        color = Ink.secondary,
    )
    Spacer(Modifier.height(8.dp))
    OutlinedTextField(
        value = typed,
        onValueChange = { typed = it },
        label = { Text("Console write token") },
        singleLine = true,
        visualTransformation = PasswordVisualTransformation(),
        modifier = Modifier.fillMaxWidth(),
        textStyle = Kind.body,
        // The caret and the focus ring ship as Material defaults and belong to
        // no design system until they are named.
        colors = OutlinedTextFieldDefaults.colors(
            focusedBorderColor = Ink.live,
            unfocusedBorderColor = Ink.rule,
            cursorColor = Ink.live,
            focusedLabelColor = Ink.live,
            unfocusedLabelColor = Ink.secondary,
            focusedTextColor = Ink.primary,
            unfocusedTextColor = Ink.primary,
        ),
    )
    Spacer(Modifier.height(8.dp))
    OutlinedButton(
        onClick = {
            onSaveToken(typed)
            typed = ""
        },
        colors = ButtonDefaults.outlinedButtonColors(contentColor = Ink.live),
    ) {
        // Saving an empty field is how a token is REMOVED, which is the only
        // way to stop a phone announcing without uninstalling the app.
        Text(if (typed.isBlank()) "Clear the stored token" else "Save the token")
    }
}
